### PR TITLE
I1307 actively manage subcard 1

### PR DESCRIPTION
**Pull request by Serhiy91**

### CODE REVIEW

File: app/imports/client/react/components/ActivleyManage/ActivelyManageSubcard.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  }
+`;
+
+const ActivelyManageSubcard = () => (
+  <Subcard>
+    <SubcardHeader>
+      <CardTitle>
+        Actively Manage
+      </CardTitle>
+      <SubcardSubtitle className="text-muted">
```
Comment (Dartv) [inline app/imports/client/react/components/ActivleyManage/ActivelyManageSubcard.js]: You can use `reactstrap`'s `FormText` to achieve the same result

Comment (Serhiy91) [reply app/imports/client/react/components/ActivleyManage/ActivelyManageSubcard.js]: I fixed it here https://github.com/Pliohub/Plio/pull/1406

File: app/imports/api/constants.js [hunk 2]
```diff
@@ -273,6 +273,12 @@ export const TransitionTimeouts = {
 
 export const TransitionBaseActiveClass = 'in';
 
+export const ProplemStatusTypes = {
```
Comment (Dartv) [inline app/imports/api/constants.js]:
```suggestion
export const ProblemStatusTypes = {
```

File: app/imports/client/graphql/Query/NonconformityList.graphql [hunk 3]
```diff
@@ -0,0 +1,8 @@
+query NonconformityList($organizationId: ID!) {
+  nonconformities(organizationId: $organizationId, type: "non-conformity") {
```
Comment (Dartv) [inline app/imports/client/graphql/Query/NonconformityList.graphql]: Maybe it's better to create a js file and insert ProblemTypes.NON_CONFORMITY instead?
You can use graphql-tag for that.

File: app/imports/client/react/canvas/components/ActivleyManage/ActivelyManageItem.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+const ActivelyManageItem = ({
+  isOpen,
+  toggle,
+  label,
+  children,
+  initialValues,
+  onSubmit,
+}) => (
+  <Fragment>
+    <ListGroupItem tag="a" onClick={toggle}>
```
Comment (Dartv) [inline app/imports/client/react/canvas/components/ActivleyManage/ActivelyManageItem.js]: Why `a`? It's not a link. You can use `Button` with `color="link"` instead

**Comment by Dartv** [comment]
Huge lag spike when opening Actively manage subcard:

Needs optimization

File: app/imports/client/react/canvas/components/ActivleyManage/ActivelyManageItem.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      Add a <strong>{label}</strong>
+      <Icon name="question-circle" />
+    </ListGroupItem>
+    <EntityForm {...{ initialValues, onSubmit }}>
+      {({ handleSubmit }) => (
+        <Form onSubmit={handleSubmit}>
+          <EntityCard {...{ isOpen }} onDelete={toggle}>
```
Comment (Dartv) [inline app/imports/client/react/canvas/components/ActivleyManage/ActivelyManageItem.js]: We need to rename this button's text to something

File: app/imports/client/react/forms/components/SubcardField.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+import { Field } from 'react-final-form';
+
+import SubcardFieldAdapter from './SubcardFieldAdapter';
+
+const SubcardField = props => (
+  <Field
```
Comment (Dartv) [inline app/imports/client/react/forms/components/SubcardField.js]: `Field` also accepts `subscription` prop. For a huge field like this we need optimization

File: app/imports/client/react/goals/components/GoalsSubcard.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+import GoalSubcard from './GoalSubcard';
+
+const GoalsSubcard = ({
+  organizationId,
+  goals,
+  onLink,
+  onUnlink,
+  canEditGoals,
+  entityName,
+  value: goalIds,
```
Comment (Dartv) [inline app/imports/client/react/goals/components/GoalsSubcard.js]: Why do you need both `goals` and `goalIds`?

File: app/imports/client/react/goals/containers/GoalEditContainer.js [hunk 8]
```diff
@@ -63,15 +63,23 @@ const getRefetchQueries = () => [
 ];
 
 const GoalEditContainer = ({
-  goalId,
+  goal: _goal = null,
+  goalId = _goal && _goal._id,
```
Comment (Dartv) [inline app/imports/client/react/goals/containers/GoalEditContainer.js]: why is this needed? I didn't need to default milestoneId with milestones

File: app/imports/client/react/goals/containers/GoalEditContainer.js [hunk 9]
```diff
@@ -99,33 +108,43 @@ const GoalEditContainer = ({
...
+              initialValues: getInitialValues({ ...goal, ...completeGoal }),
+            })}
           />,
           <Mutation
             mutation={Mutations.UNDO_GOAL_COMPLETION}
             refetchQueries={getRefetchQueries}
             children={noop}
+            onCompleted={
+              ({ undoGoalCompletion }) => setState({
+                goal: { ...goal, ...undoGoalCompletion },
```
Comment (Dartv) [inline app/imports/client/react/goals/containers/GoalEditContainer.js]: Actually, I think you don't need to update state after mutation. Try passing `notifyOnNetworkStatusChange` prop to `Query` first. If it doesn't work then use `merge` instead

File: app/imports/client/react/goals/containers/GoalEditContainer.js [hunk 10]
```diff
@@ -99,33 +108,43 @@ const GoalEditContainer = ({
...
           ...props,
-          loading,
           error,
           organizationId,
           isOpen,
           toggle,
           initialValues,
           goal,
           canEditGoals,
+          loading: !_goal && loading,
```
Comment (Dartv) [inline app/imports/client/react/goals/containers/GoalEditContainer.js]: why?

File: app/imports/client/react/goals/containers/GoalEditContainer.js [hunk 11]
```diff
@@ -203,10 +222,42 @@ const GoalEditContainer = ({
...
-          }).then(toggle) : undefined,
+          onDelete: canEditGoals || onUnlink ? () => {
+            if (!onUnlink) {
+              return onDelete({
+                mutate: deleteGoal,
+                ownProps: { organizationId, goal },
+              }).then(toggle);
+            }
+            return swal.promise({
+              title: canEditGoals && 'Choose an action',
```
Comment (Dartv) [inline app/imports/client/react/goals/containers/GoalEditContainer.js]: If you can't edit goals, you shouldn't be able to delete them either

File: app/imports/client/react/helpers/withToggle.js [hunk 12]
```diff
@@ -1,9 +1,9 @@
 import { branch } from 'recompose';
-import { prop, identity } from 'ramda';
+import { isNil, identity } from 'ramda';
 import withStateToggle from './withStateToggle';
 
 const withToggle = (defaultValue = false) => branch(
-  prop('toggle'),
+  ({ toggle, isOpen }) => toggle || !isNil(isOpen),
```
Comment (Dartv) [inline app/imports/client/react/helpers/withToggle.js]: what's the reasoning? This could break our code

File: app/imports/client/react/noncomformities/components/NonconformitiesSubcard.js [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+const NonconformitiesSubcard = ({
+  organizationId,
+  nonconformities,
+  onLink,
+  type,
+  guidelines,
+  onUnlink,
+  entityName,
+  value: nonconformityIds,
```
Comment (Dartv) [inline app/imports/client/react/noncomformities/components/NonconformitiesSubcard.js]: Why do you need both nonconformities and nonconformityIds there?

**Comment by Dartv** [comment]
You should refactor according to `SelectInput` component changes from https://github.com/Pliohub/Plio/pull/1490 and also date changes from https://github.com/Pliohub/Plio/pull/1480

**Comment by Dartv** [comment]
The top form lags like crazy because of Actively manage subcard. Even changing title freezes the form

File: app/imports/client/react/noncomformities/containers/NonconformityEditContainer.js [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+    'departments',
+    'status',
+    'sequentialId',
+    'cost',
+    'ref',
+  ]),
+);
+
+const NonconformityEditContainer = ({
+  nonconformity: nonconformityDoc,
```
Comment (Dartv) [inline app/imports/client/react/noncomformities/containers/NonconformityEditContainer.js]: why not follow other docs edit containers? It will backfire later on.

File: app/imports/client/react/noncomformities/containers/NonconformityEditContainer.js [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+              form.reset(currentValues);
+              throw err;
+            });
+          },
+          onDelete: () => {
+            const nonconformityName = type === ProblemTypes.NON_CONFORMITY ?
+              'nonconformity' : 'potential gain';
+            return swal.promise({
+              title: 'Choose an action',
+              text: `Do you wish to unlink the ${nonconformityName} "${nonconformity.title}" ` +
```
Comment (Dartv) [inline app/imports/client/react/noncomformities/containers/NonconformityEditContainer.js]: This will not work later on when we'll need to use this as top form container

File: app/imports/client/react/risks/containers/RiskAddContainer.js [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+import { validateRisk } from '../../../validation';
+import { Composer, renderComponent } from '../../helpers';
+import { swal } from '../../../util';
+
+const RiskAddContainer = ({
+  organizationId,
+  isOpen,
+  toggle,
+  onLink,
+  entityId,
```
Comment (Dartv) [inline app/imports/client/react/risks/containers/RiskAddContainer.js]: What do you need `entityId` there for?

File: app/imports/client/react/standards/components/StandardSectionSelectInput.js [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+        placeholder="Standards section"
+        promptTextCreator={selectedLabel => `Add "${selectedLabel}" section`}
+        loadOptions={query => query({
+          query: Queries.STANDARD_SECTION_LIST,
+          variables: { organizationId },
+        }).then(({ data: { standardSections: { standardSections = [] } = {} } }) => ({
+          options: getOptions(standardSections),
+        })).catch(swal.error)}
+        onNewOptionClick={({ value: title }, callback) => {
+          if (!title) return null;
```
Comment (Dartv) [inline app/imports/client/react/standards/components/StandardSectionSelectInput.js]: what's this for?

File: app/imports/client/react/standards/components/StandardsSubcard.js [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+import NewStandardCard from './NewStandardCard';
+import StandardSubcard from './StandardSubcard';
+
+const StandardsSubcard = ({
+  organizationId,
+  standards,
+  onLink,
+  onUnlink,
+  entityName,
+  value: standardIds,
```
Comment (Dartv) [inline app/imports/client/react/standards/components/StandardsSubcard.js]: ?

File: app/imports/client/react/standards/containers/StandardEditContainer.js [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+const enhance = composeWithTracker(
+  ({ standard: { _id }, isOpen }, onData) => {
+    if (isOpen) {
+      Meteor.subscribe(
+        'sourceFilesByDocument',
+        { _id, documentType: DocumentTypes.STANDARD },
+        { onStop: error => error && swal.error(error, 'Files subscription error') },
+      );
+    }
+    onData(null, {});
```
Comment (Dartv) [inline app/imports/client/react/standards/containers/StandardEditContainer.js]: Wait, how does this work if you don't pass any props forward?

File: app/imports/client/react/standards/containers/StandardEditContainer.js [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+    'issueNumber',
+    'departments',
+  ]),
+);
+
+const enhance = composeWithTracker(
+  ({ standard: { _id }, isOpen }, onData) => {
+    if (isOpen) {
+      Meteor.subscribe(
+        'sourceFilesByDocument',
```
Comment (Dartv) [inline app/imports/client/react/standards/containers/StandardEditContainer.js]: You don't need to fetch files there?

File: app/imports/client/react/standards/containers/StandardEditContainer.js [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+        { onStop: error => error && swal.error(error, 'Files subscription error') },
+      );
+    }
+    onData(null, {});
+  },
+  { propsToWatch: ['isOpen'] },
+);
+
+const StandardEditContainer = ({
+  standard: standardDoc,
```
Comment (Dartv) [inline app/imports/client/react/standards/containers/StandardEditContainer.js]: Same as for risks. This will be a problem in the future when we'll need to use this container for the top form

File: app/packages/meteor-bootstrap-sweetalert/dev/modules/handle-click.js [hunk 22]
```diff
@@ -65,14 +78,12 @@ var handleConfirm = function(modal, params) {
   }
 
   params.doneFunction(callbackValue);
+  closeModal(params);
```
Comment (Dartv) [inline app/packages/meteor-bootstrap-sweetalert/dev/modules/handle-click.js]: Reasoning for this?

File: app/packages/meteor-bootstrap-sweetalert/dev/modules/set-params.js [hunk 23]
```diff
@@ -26,6 +26,7 @@ var setParameters = function(params) {
   var $text = modal.querySelector('p');
   var $cancelBtn = modal.querySelector('button.cancel');
   var $confirmBtn = modal.querySelector('button.confirm');
+  var $midBtn = modal.querySelector('button.mid-button');
```
Comment (Dartv) [inline app/packages/meteor-bootstrap-sweetalert/dev/modules/set-params.js]: Maybe it's better to call this `extra-button`?

File: graphql-server/imports/api/apollo/resolvers/Mutation/Nonconformity/createNonconformity.js [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+  insertAfterware,
+} from '../../../../../share/middleware';
+
+export const resolver = async (root, args, context) =>
+  context.services.NonConformityService.insert(args, context);
+
+export default applyMiddleware(
+  flattenInput(),
+  checkLoggedIn(),
+  checkOrgMembership(),
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/resolvers/Mutation/Nonconformity/createNonconformity.js]: You should check `owner` and `originator` access too

File: graphql-server/imports/api/apollo/resolvers/Mutation/Nonconformity/updateNonconformity.js [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+  branch,
+  checkOrgMembership,
+  checkStandardsAccess,
+  nonconformityUpdateAfterware,
+} from '../../../../../share/middleware';
+
+export const resolver = async (root, args, context) =>
+  context.services.NonConformityService.update(args, context);
+
+export default applyMiddleware(
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/resolvers/Mutation/Nonconformity/updateNonconformity.js]: Where is departments access check?
Where is originator access check?

File: graphql-server/imports/api/apollo/resolvers/Mutation/Standard/createStandard.js [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+  checkLoggedIn,
+  checkOrgMembership,
+  flattenInput,
+  insertAfterware,
+} from '../../../../../share/middleware';
+
+export const resolver = async (root, args, { services: { StandardService } }) =>
+  StandardService.insert(args);
+
+export default applyMiddleware(
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/resolvers/Mutation/Standard/createStandard.js]: Where's standard section access check?
Where's standard type access check?
Where's owner access check?
Where's source file cess check?

File: graphql-server/imports/api/apollo/resolvers/Mutation/Standard/updateStandard.js [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+  branch,
+  checkOrgMembership,
+  checkDepartmentsAccess,
+  standardUpdateAfterware,
+} from '../../../../../share/middleware';
+
+export const resolver = async (root, args, context) =>
+  context.services.StandardService.update(args, context);
+
+export default applyMiddleware(
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/resolvers/Mutation/Standard/updateStandard.js]: Where's standard section access check?
Where's standard type access check?
Where's source files access check?

File: graphql-server/imports/api/apollo/resolvers/Query/Nonconformity/nonconformities.js [hunk 28]
```diff
@@ -0,0 +1,24 @@
...
+
+export const resolver = async (
+  root,
+  { organizationId, type },
+  { collections: { NonConformities } },
+) => {
+  const cursor = NonConformities.find({
+    organizationId,
+    type,
+    isDeleted: false,
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/resolvers/Query/Nonconformity/nonconformities.js]: `isDeleted` should be an optional argument

File: graphql-server/imports/api/apollo/resolvers/Types/util.js [hunk 29]
```diff
@@ -76,3 +80,64 @@ export const resolveCustomerElementStatus = async (root, args, context) => {
 
   return isMatched ? CustomerElementStatuses.MATCHED : CustomerElementStatuses.UNMATCHED;
 };
+
+export const loadGoalsByIds = async (root, args, context) => {
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/resolvers/Types/util.js]: Why use different naming if we already established it? Why not call it `resolveGoals`? These new resolvers should be configurable too. Look at how relation resolvers are done.

File: graphql-server/imports/api/apollo/resolvers/Types/util.js [hunk 30]
```diff
@@ -76,3 +80,64 @@ export const resolveCustomerElementStatus = async (root, args, context) => {
...
+  const { goalIds = [] } = root;
+  const { loaders: { Goal: { byQuery } } } = context;
+
+  return byQuery.load({
+    _id: { $in: goalIds },
+    isDeleted: false,
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/resolvers/Types/util.js]: We don't need to show completed goals too I think?

File: graphql-server/imports/api/apollo/schema/Interface/CanvasEntity.graphql [hunk 31]
```diff
@@ -5,4 +5,5 @@ interface CanvasEntity {
   color: String!
   notes: String
   notify: [User]
+  organization: Organization
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/schema/Interface/CanvasEntity.graphql]: organization should be required

File: graphql-server/imports/api/apollo/schema/Types/Nonconformity.graphql [hunk 32]
```diff
@@ -0,0 +1,36 @@
+type HelpDeskRef {
+  text: String
+  url: String
+}
+
+type Nonconformity implements BaseEntity {
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/schema/Types/Nonconformity.graphql]: Also, PlioEntity

File: graphql-server/imports/api/apollo/schema/Types/Nonconformity.graphql [hunk 33]
```diff
@@ -0,0 +1,36 @@
...
+  _id: ID!
+  organization: Organization!
+  createdAt: DateTime!
+  createdBy: User!
+  updatedAt: DateTime
+  updatedBy: User
+  title: String!
+  description: String,
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/schema/Types/Nonconformity.graphql]: typo

File: graphql-server/imports/api/apollo/schema/Types/Standard.graphql [hunk 34]
```diff
@@ -17,13 +24,14 @@ type Standard implements BaseEntity, PlioEntity {
   deletedBy: User
   notify: [User]
   status: String!
-  # departments: [Department]
+  departments: [Department]
   files: [File]
-  # type: StandardType
-  # section: StandardBookSection
+  typeId: String
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/schema/Types/Standard.graphql]: We don't need this

File: share/middleware/goal/checkGoalsAccess.js [hunk 35]
```diff
@@ -0,0 +1,8 @@
+import { checkDocsAccess } from '../document';
+
+export default (config = () => ({})) =>
```
Comment (Dartv) [inline share/middleware/goal/checkGoalsAccess.js]: use `branch` there

File: share/middleware/risk/checkRisksAccess.js [hunk 36]
```diff
@@ -0,0 +1,8 @@
+import { checkDocsAccess } from '../document';
+
+export default (config = () => ({})) =>
```
Comment (Dartv) [inline share/middleware/risk/checkRisksAccess.js]: use `branch` there

File: app/imports/client/react/noncomformities/helpers.js [hunk 37]
```diff
@@ -0,0 +1,15 @@
+import { StatusColors, ProplemStatusTypes } from '../../../api/constants';
+import { getProblemStatusColor } from '../../../api/problems/helpers';
+
+export const getStatusColor = (status) => {
```
[the remark this answers is not in the document]
Comment (Serhiy91) [reply app/imports/client/react/noncomformities/helpers.js]: @Dartv Otherwise, New nonconformity form will be different then old form.

File: app/imports/client/react/helpers/withToggle.js (hunk 12 above)
Comment (Serhiy91) [reply app/imports/client/react/helpers/withToggle.js]: It is just additional checking of `isOpen`. It prevents overwriting of `isOpen`.

File: app/imports/client/react/standards/containers/StandardEditContainer.js (hunk 20 above)
Comment (Serhiy91) [reply app/imports/client/react/standards/containers/StandardEditContainer.js]: No, ESSources template fetches files

File: app/imports/api/standards-book-sections/standards-book-section-service.js [hunk 38]
```diff
@@ -1,24 +1 @@
...
-      $set: { title },
-    });
-  },
-
-  remove({ _id }) {
-    return this.collection.remove({ _id });
-  },
-
-};
+export { default } from '../../share/services/standard-service';
```
Comment (Dartv) [inline app/imports/api/standards-book-sections/standards-book-section-service.js]: But this is Standard book sections service, not standard service

File: app/imports/client/react/canvas/components/ActivleyManage/RiskActivelyManageItem.js [hunk 39]
```diff
@@ -5,30 +5,20 @@ import { EntityManagerItem } from '../../../components';
 import { NewRiskCard, RiskAddContainer } from '../../../risks';
 import ActivelyManageItem from './ActivelyManageItem';
 
-const RiskActivelyManageItem = ({
-  organizationId,
-  linkedTo,
-  onLink,
-  ...rest
-}) => (
+const RiskActivelyManageItem = ({ organizationId, onLink, ...rest }) => (
```
Comment (Dartv) [inline app/imports/client/react/canvas/components/ActivleyManage/RiskActivelyManageItem.js]: We don't have TypeScript so it's very hard to understand what props can be passed in. To avoid confusion, it's better to pass through all expected props, even if they're not used directly in this component. Exceptions, of course, are reusable components and components that use `renderComponent`, etc...

File: app/imports/client/react/canvas/components/CanvasLessonsSubcard.js [hunk 40]
```diff
@@ -3,23 +3,9 @@ import PropTypes from 'prop-types';
...
 
-const CanvasLessonsSubcard = ({
-  organizationId,
-  lessons,
-  linkedTo,
-  documentType,
-  refetchQuery,
-  entityName,
-}) => (
+const CanvasLessonsSubcard = ({ linkedTo, ...rest }) => (
```
Comment (Dartv) [inline app/imports/client/react/canvas/components/CanvasLessonsSubcard.js]: same as https://github.com/Pliohub/Plio/pull/1382/files/331db5de0d038c5bb4108df344f6f6d13b35e31d..d47a6963e8831572e9279b9cd220eca4e0945d7e#r229704381

File: app/imports/client/react/canvas/components/CanvasRisksSubcard.js [hunk 41]
```diff
@@ -1,29 +1,10 @@
 import React from 'react';
-import PropTypes from 'prop-types';
 
-import { Magnitudes } from '../../components';
 import { RisksSubcard, RiskAddContainer } from '../../risks';
 
-const CanvasRisksSubcard = ({
```
Comment (Dartv) [inline app/imports/client/react/canvas/components/CanvasRisksSubcard.js]: https://github.com/Pliohub/Plio/pull/1382/files/331db5de0d038c5bb4108df344f6f6d13b35e31d..d47a6963e8831572e9279b9cd220eca4e0945d7e#r229704381

File: app/imports/client/react/forms/components/GoalSelectInput.js [hunk 42]
```diff
@@ -0,0 +1,27 @@
...
+import ApolloSelectInputField from './ApolloSelectInputField';
+
+const GoalSelectInput = ({ organizationId, goalIds = [], ...props }) => (
+  <ApolloSelectInputField
+    {...props}
+    loadOptions={query => query({
+      query: Query.GOAL_LIST,
+      variables: { organizationId },
+    }).then(({ data: { goals: { goals } } }) => ({
+      options: mapRejectedEntitiesByIdsToOptions(goalIds, goals),
```
Comment (Dartv) [inline app/imports/client/react/forms/components/GoalSelectInput.js]: We'll need to make this customizable, but in the different PRs

File: app/imports/client/react/forms/components/NonconformitySelectInput.js [hunk 43]
```diff
@@ -18,7 +18,7 @@ const NonconformitySelectInput = ({
     loadOptions={query => query({
       query: type === ProblemTypes.NON_CONFORMITY ?
         Query.NONCONFORMITIES_LIST : Query.POTENTIAL_GAIN_LIST,
-      variables: { organizationId, type },
+      variables: { organizationId },
```
Comment (Dartv) [inline app/imports/client/react/forms/components/NonconformitySelectInput.js]: In the future, we'll split Nonconformities and potential gains into multiple collections, so we should create different select for each one.

File: app/imports/client/react/forms/components/NonconformitySelectInput.js [hunk 44]
```diff
@@ -18,7 +18,7 @@ const NonconformitySelectInput = ({
     loadOptions={query => query({
       query: type === ProblemTypes.NON_CONFORMITY ?
         Query.NONCONFORMITIES_LIST : Query.POTENTIAL_GAIN_LIST,
```
Comment (Dartv) [inline app/imports/client/react/forms/components/NonconformitySelectInput.js]: Why is query called `NONCONFORMITIES_LIST` when it should be called `NONCONFORMITY_LIST`?

File: app/imports/client/react/goals/containers/GoalEditContainer.js [hunk 45]
```diff
@@ -93,6 +92,7 @@ const GoalEditContainer = ({
               initialValues: getInitialValues(getGoal(data)),
               goal: getGoal(data),
             })}
+            notifyOnNetworkStatusChange
```
Comment (Dartv) [inline app/imports/client/react/goals/containers/GoalEditContainer.js]: Why do you need `notifyOnNetworkStatusChange` if you're updating state after mutation manually? I asked to try this out without updating state after mutation, and if it works, delete all `onCompleted` calls from mutations

File: app/imports/client/react/goals/containers/GoalSubcardContainer.js [hunk 46]
```diff
@@ -0,0 +1,62 @@
...
+  </Mutation>
+);
+
+GoalSubcardContainer.propTypes = {
+  organizationId: PropTypes.string.isRequired,
+  canEditGoals: PropTypes.bool,
+  goal: PropTypes.object.isRequired,
+  toggle: PropTypes.func.isRequired,
+  entityName: PropTypes.string,
+  onUnlink: PropTypes.func,
```
Comment (Dartv) [inline app/imports/client/react/goals/containers/GoalSubcardContainer.js]: You're declaring onUnlink as not required, but if it's not provided the error will be thrown by looking at your code

File: app/imports/client/react/standards/containers/StandardAddContainer.js [hunk 47]
```diff
@@ -107,14 +107,20 @@ const StandardAddContainer = ({
...
+        let insertFilePromise;
         if (source1.type === 'attachment' && file) {
-          fileId = await insertFile.call({
-            name: file.name,
-            extension: file.name.split('.').pop().toLowerCase(),
-            organizationId,
+          insertFilePromise = new Promise((resolve, reject) => {
+            insertFile.call({
+              name: file.name,
+              extension: file.name.split('.').pop().toLowerCase(),
```
Comment (Dartv) [inline app/imports/client/react/standards/containers/StandardAddContainer.js]: shouldn't this be done in server-side code to ensure security?

File: app/imports/client/react/standards/containers/StandardAddContainer.js [hunk 48]
```diff
@@ -107,14 +107,20 @@ const StandardAddContainer = ({
...
+              organizationId,
+            }, (error, fileId) => {
+              if (error) reject(error);
+              resolve(fileId);
+            });
           });
         }
+        const fileId = insertFilePromise && await insertFilePromise;
```
Comment (Dartv) [inline app/imports/client/react/standards/containers/StandardAddContainer.js]: You don't handle the case when the error is thrown

Comment (Dartv) [reply app/imports/client/react/standards/containers/StandardAddContainer.js]: What will happen if it returns false? It will pass false as `id` to the method call. We don't want that.

File: graphql-server/imports/api/apollo/resolvers/Mutation/Nonconformity/createNonconformity.js [hunk 49]
```diff
@@ -13,7 +14,20 @@ export const resolver = async (root, args, context) =>
 export default applyMiddleware(
   flattenInput(),
   checkLoggedIn(),
-  checkOrgMembership(),
+  branch(
+    (root, args) => args.ownerId,
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/resolvers/Mutation/Nonconformity/createNonconformity.js]: But ownerId and originatorId are required in `create` mutation

File: graphql-server/imports/api/apollo/resolvers/Mutation/Standard/createStandard.js [hunk 50]
```diff
@@ -12,7 +16,22 @@ export const resolver = async (root, args, { services: { StandardService } }) =>
...
-  checkOrgMembership(),
+  branch(
+    (root, args) => args.owner,
+    checkOrgMembership((root, { organizationId, owner }) => ({
+      organizationId,
+      userId: owner,
+    })),
+  ),
+  branch(
+    (root, args) => args.source1.fileId,
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/resolvers/Mutation/Standard/createStandard.js]: this will throw some weird error to the user if `source1` is null or undefined. So you should handle it too

File: graphql-server/imports/api/apollo/resolvers/Mutation/Standard/updateStandard.js [hunk 51]
```diff
@@ -1,4 +1,4 @@
-import { applyMiddleware } from 'plio-util';
+import { applyMiddleware, getAllStandardFileIds } from 'plio-util';
```
Comment (Dartv) [inline graphql-server/imports/api/apollo/resolvers/Mutation/Standard/updateStandard.js]: Don't put it inside plio util

Comment (Serhiy91) [reply graphql-server/imports/api/apollo/resolvers/Mutation/Standard/updateStandard.js]: I have just reused it. It is using in other places.

File: app/imports/client/react/standards/containers/StandardAddContainer.js (hunk 47 above)
Comment (Serhiy91) [reply app/imports/client/react/standards/containers/StandardAddContainer.js]: It causes a lot of errors when it is moved in server-side. This needs a separate task.

**Comment by Dartv** [comment]
Please, describe what changes have been made

**Comment by Serhiy91** [comment]
Following code/components were added:
- ActivelyManageSubcard which consist of newly created components: ActivelyManageItem, GoalActivelyManageItem, LessonActivelyManageItem, NonconformityActivelyManageItem, PotentialGainActivelyManageItem, RiskActivelyManageItem, StandardActivelyManageItem.
- GoalsSubcard component which consist of newly created components: GoalSubcard, NewGoalCard, GoalAddContainer, GoalSubcardContainer.
- NonconformitiesSubcard component which consist of newly created components: NonconformitySubcard, NewNonconformityCard, NewPotentialGainCard, NonconformityAddForm, NonconformityAddFormWrapper, NonconformityEditForm, NonconformityAddContainer, NonconformityEditContainer, NonconformitySubcardContainer.
- StandardsSubcard component which consist of newly created components:  StandardSubcard, NewStandardCard, StandardAddForm, StandardAddFormWrapper, StandardEditForm, StandardAddContainer, StandardEditContainer, StandardSubcardContainer.
- Reusable form components: CreateSourceField, EditSourceField, DepartmentsCreatableField, NonconformitySelectInput, PotentialGainSelectInput, StandardSectionSelectInput, StandardStatusField, StandardTypeSelectInput, EntitiesField, NewExistingSwitchField.
- withExtraAction method into swal.
- KeyPartnerEditModal was extended via ActivelyManageSubcard, GoalsSubcard, StandardsSubcard, NonconformitiesSubcard, PotantialGainsSubcard, RisksSubcard, LessonsSubcard.
- graphql server code of nonconformity/potential gain
- graphql server code of standard
- graphql server code of standard section
- graphql server code of standard type
